File: Chains/GrabEdge.cpp
#include <cmath>

#include "GrabEdge.h"
#include "TransitionHelper.h"

void GrabEdge::PressButtons()
{
    if(m_isInFastfall)
    {
        m_controller->emptyInput();
        return;
    }

    //We can't run if we're in knee bend or shine. We're in a wavedash so let's just accept that
    if(m_state->m_memory->player_two_action == KNEE_BEND ||
        m_state->m_memory->player_two_action == DOWN_B_GROUND)
    {
        m_isInWavedash = true;
    }

    //If we're far away from the edge, then dash at the edge
    if(!m_isInWavedash && (std::abs(m_state->m_memory->player_two_x) < m_state->getStageEdgeGroundPosition() - 13))
    {
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, m_isLeftEdge ? 0 : 1, .5);
        return;
    }

    m_isInWavedash = true;

    //Fastfall once in the air
    if((m_state->m_memory->player_two_action == FALLING) &&
        !m_isInFastfall)
    {
        if(m_state->m_memory->player_two_speed_y_self < 0)
        {
            m_isInFastfall = true;
            m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, 0);
            return;
        }
        else
        {
            m_controller->emptyInput();
            return;
        }

    }

    double distanceFromEdge = m_state->getStageEdgeGroundPosition() - std::abs(m_state->m_memory->player_two_x);

    //Dash Backwards if we're facing towards the edge and are in a state where we can dash
    // Or if we're too close to safely wavedash back
    if(m_isLeftEdge != m_state->m_memory->player_two_facing ||
        distanceFromEdge < 3)
    {
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, m_isLeftEdge ? 1 : 0, .5);
        return;
    }

    //Once we're dashing away from the edge, jump
    if(m_state->m_memory->player_two_action == DASHING ||
        m_state->m_memory->player_two_action == STANDING ||
        m_state->m_memory->player_two_action == DOWN_B_GROUND)
    {
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, .5, .5);
        m_controller->pressButton(Controller::BUTTON_Y);
        return;
    }

    //Just hang out and do nothing while initial knee bending
    if(m_state->m_memory->player_two_action == KNEE_BEND &&
        m_state->m_memory->player_two_action_frame < 3)
    {
        m_controller->emptyInput();
        return;
    }

    //Apparently you can be temporarily standing a little past the normal edge of the stage
    double edgeDistance = std::abs(m_state->getStageEdgeGroundPosition() + .5 - std::abs(m_state->m_memory->player_two_x));
    bool slidingTowardEdge = (m_state->m_memory->player_two_speed_ground_x_self > 0) != m_isLeftEdge;

    //If we're about to slide off next frame. Just let it happen, don't air dodge
    if(slidingTowardEdge &&
        std::abs(m_state->m_memory->player_two_speed_ground_x_self) > edgeDistance)
    {
        m_controller->emptyInput();
        return;
    }

    //Once we're in the air, airdodge backwards to the edge
    if((!m_state->m_memory->player_two_on_ground && m_isInWavedash) ||
        (m_state->m_memory->player_two_action == KNEE_BEND &&
        m_state->m_memory->player_two_action_frame >= 3))
    {
        m_controller->pressButton(Controller::BUTTON_L);
        m_controller->tiltAnalog(Controller::BUTTON_MAIN, m_isLeftEdge ? .2 : .8, .2);
        return;
    }

    //Just hang out and do nothing while wave landing
    if(m_state->m_memory->player_two_action == WAVEDASH_SLIDE)
    {
        m_controller->emptyInput();
        return;
    }

    m_controller->emptyInput();
}

bool GrabEdge::IsInterruptible()
{
    bool facingOut = m_state->m_memory->player_two_facing == m_isLeftEdge;

    //If we need to fastfall and haven't yet, don't leave
    if((m_state->m_memory->player_two_action == FALLING) && !m_isInFastfall)
    {
        return false;
    }

    //If we're in this state, something horrible has happened. We won't grab the edge. Abort
    if(facingOut && !m_state->m_memory->player_two_on_ground)
    {
        return true;
    }

    if(!m_isInWavedash)
    {
        return true;
    }
    if(m_state->m_memory->player_two_action == EDGE_HANGING)
    {
        return true;
    }
    //Don't permanently get stuck here
    uint frame = m_state->m_memory->frame - m_startingFrame;
    if(frame > 60)
    {
        return true;
    }
    return false;
}

GrabEdge::GrabEdge()
{
    //Quick variable to tell us which edge to grab
    if(m_state->m_memory->player_one_x > 0)
    {
        m_isLeftEdge = false;
    }
    else
    {
        m_isLeftEdge = true;
    }
    m_startingFrame = m_state->m_memory->frame;
    m_isInWavedash = false;
    m_isInFastfall = false;
}

GrabEdge::~GrabEdge()
{
}

File: Chains/grabedge.py
import melee
from melee.enums import Action, Button
from Chains.chain import Chain

# Grab the edge
class Grabedge(Chain):
    def __init__(self, wavedash=True):
        self.wavedash = wavedash

    def step(self, gamestate, smashbot_state, opponent_state):
        controller = self.controller

        # Moved this here from constructor.
        #   It should be fine, but let's keep an eye out for if this breaks
        edge_x = melee.stages.EDGE_GROUND_POSITION[gamestate.stage]
        if opponent_state.x < 0:
            edge_x = -edge_x
        edgedistance = abs(edge_x - smashbot_state.x)
        if edgedistance > 15:
            self.wavedash = False
        if edgedistance < 2:
            self.wavedash = False

        # Where is the edge that we're going to?
        edge_x = melee.stages.EDGE_GROUND_POSITION[gamestate.stage]
        if opponent_state.x < 0:
            edge_x = -edge_x

        # If we're on the edge, then we're done here, end the chain
        if smashbot_state.action in [Action.EDGE_HANGING, Action.EDGE_CATCHING]:
            self.interruptible = True
            controller.empty_input()
            return

        # If we're in spotdodge, do nothing
        if smashbot_state.action == Action.SPOTDODGE:
            self.interruptible = True
            controller.empty_input()
            return

        # If we're stuck wavedashing, just hang out and do nothing
        if smashbot_state.action == Action.LANDING_SPECIAL:
            self.interruptible = False
            controller.empty_input()
            return

        #If we're walking, stop for a frame
        #Also, if we're shielding, don't try to dash. We will accidentally roll
        if smashbot_state.action == Action.WALK_SLOW or \
            smashbot_state.action == Action.WALK_MIDDLE or \
            smashbot_state.action == Action.WALK_FAST or \
            smashbot_state.action == Action.SHIELD_START or \
            smashbot_state.action == Action.SHIELD_REFLECT or \
            smashbot_state.action == Action.SHIELD:
                self.interruptible = True
                controller.empty_input()
                return

        facinginwards = smashbot_state.facing == (smashbot_state.x < 0)
        if smashbot_state.action == Action.TURNING and smashbot_state.action_frame == 1:
            facinginwards = not facinginwards

        edgedistance = abs(edge_x - smashbot_state.x)
        turnspeed = abs(smashbot_state.speed_ground_x_self)
        # If we turn right now, what will our speed be?
        if smashbot_state.action == Action.DASHING:
            turnspeed = (abs(smashbot_state.speed_ground_x_self) - 0.32) / 4
        slidedistance = self.framedata.slide_distance(smashbot_state, turnspeed, 7)
        closetoedge = edgedistance < slidedistance

        # Do a wavedash off
        if self.wavedash and not smashbot_state.off_stage:
            self.interruptible = False
            if smashbot_state.action == Action.KNEE_BEND and smashbot_state.action_frame >= 3:
                controller.press_button(Button.BUTTON_L)
                controller.release_button(Button.BUTTON_Y)
                controller.tilt_analog(melee.Button.BUTTON_MAIN, int(not smashbot_state.facing), 0.2)
                return
            if facinginwards:
                if controller.prev.button[Button.BUTTON_Y]:
                    controller.release_button(Button.BUTTON_Y)
                    return
                else:
                    controller.press_button(Button.BUTTON_Y)
                    return
            else:
                # Dash inwards to turn
                controller.tilt_analog(melee.Button.BUTTON_MAIN, int(not smashbot_state.facing), 0.5)
                return

        # This is actually shine turnaround
        if smashbot_state.action == Action.MARTH_COUNTER:
            self.interruptible = False
            controller.empty_input()
            return

        # If we're in the shine, but too high, just wait
        if smashbot_state.action in [Action.SWORD_DANCE_4_MID_AIR, Action.SWORD_DANCE_4_LOW_AIR] \
                and -10 < smashbot_state.y and edgedistance < 10:
            self.interruptible = False
            controller.empty_input()
            return

        # Fastfall, but only once
        if smashbot_state.action == Action.FALLING:
            self.interruptible = False

            # Should we shine?
            canhit = gamestate.distance < 11.8 and opponent_state.invulnerability_left == 0
            if (smashbot_state.y < -15) or canhit:
                controller.press_button(Button.BUTTON_B)
                controller.tilt_analog(melee.Button.BUTTON_MAIN, 0.5, 0)
                return

            # Fastfall speed is 3.4, but we need a little wiggle room
            if smashbot_state.speed_y_self > -3.35:
                controller.tilt_analog(melee.Button.BUTTON_MAIN, 0.5, 0)
            else:
                # DI in to the opponent
                x = 0
                if smashbot_state.x < opponent_state.x:
                    x = 1
                controller.tilt_analog(melee.Button.BUTTON_MAIN, x, 0.5)
            return

        # Shine turnaround
        if smashbot_state.action == Action.DOWN_B_STUN and not facinginwards:
            self.interruptible = False
            controller.tilt_analog(melee.Button.BUTTON_MAIN, int(not smashbot_state.facing), .5)
            return

        # Jump out of shine
        if smashbot_state.action == Action.DOWN_B_AIR and facinginwards:
            self.interruptible = False
            controller.release_button(Button.BUTTON_B)
            controller.tilt_analog(Button.BUTTON_MAIN, 0.5, 0.5)
            if controller.prev.button[Button.BUTTON_Y]:
                controller.release_button(Button.BUTTON_Y)
                return
            else:
                controller.press_button(Button.BUTTON_Y)
                return

        # Firefox to grab edge
        if smashbot_state.action == Action.JUMPING_ARIAL_FORWARD:
            # Must be between 0 and -10
            inxrange = -10 < (abs(edge_x) - abs(smashbot_state.x)) < 0
            if -15 < smashbot_state.y < -5 and inxrange:
                self.interruptible = False
                controller.press_button(Button.BUTTON_B)
                controller.tilt_analog(melee.Button.BUTTON_MAIN, 0.5, 1)
                return
            else:
                self.interruptible = False
                controller.empty_input()
                return

        # Pivot slide
        if smashbot_state.action == Action.TURNING and facinginwards and closetoedge:
            self.interruptible = False
            controller.empty_input()
            return

        # Do the turn
        if smashbot_state.action == Action.DASHING and closetoedge:
            self.interruptible = False
            controller.tilt_analog(melee.Button.BUTTON_MAIN, int(not smashbot_state.facing), .5)
            return

        #If we're starting the turn around animation, keep pressing that way or
        #   else we'll get stuck in the slow turnaround
        if smashbot_state.action == Action.TURNING and smashbot_state.action_frame == 1:
            self.interruptible = True
            return

        #Dash back, since we're about to start running
        if smashbot_state.action == Action.DASHING and smashbot_state.action_frame >= 11:
            controller.tilt_analog(melee.Button.BUTTON_MAIN, int(not smashbot_state.facing), .5)
            self.interruptible = True
            return

        #We can't dash IMMEDIATELY after landing. So just chill for a bit
        if smashbot_state.action == Action.LANDING and smashbot_state.action_frame < 2:
            self.interruptible = True
            controller.empty_input()
            return

        #Are we outside the given radius of dash dancing?
        if smashbot_state.x < edge_x:
            self.interruptible = True
            controller.tilt_analog(melee.Button.BUTTON_MAIN, 1, .5)
            return

        if smashbot_state.x > edge_x:
            self.interruptible = True
            controller.tilt_analog(melee.Button.BUTTON_MAIN, 0, .5)
            return

        #Keep running the direction we're going
        self.interruptible = True
        controller.tilt_analog(melee.Button.BUTTON_MAIN, int(not smashbot_state.facing), .5)
        return
